src/Microsoft.AspNet.Authentication/AuthenticationHandler.cs

```diff
@@ -395,11 +235,48 @@ public virtual bool ShouldConvertChallengeToForbidden()
         /// deals an authentication interaction as part of it's request flow. (like adding a response header, or
         /// changing the 401 result to 302 of a login page or external sign-in location.)
         /// </summary>
-        /// <returns></returns>
-        protected virtual Task ApplyResponseChallengeAsync()
+        /// <param name="context"></param>
+        /// <returns>True if no other handlers should be called</returns>
+        protected virtual Task<bool> HandleUnauthorizedAsync(ChallengeContext context)
         {
-            ApplyResponseChallenge();
-            return Task.FromResult(0);
+            Response.StatusCode = 401;
+            return Task.FromResult(false);
+        }
+
+        public async Task ChallengeAsync(ChallengeContext context)
+        {
+            bool handled = false;
+            ChallengeCalled = true;
```

Comment:
This one is slightly different, since we are using it as a guard around HandleAutomaticChallengeIfNeeded, I will rename the other two to Accepted and leave this one called